### PR TITLE
[SPARK-43128][CONNECT][SS] Make `recentProgress` and `lastProgress` return `StreamingQueryProgress` consistent with the native Scala Api

### PR DESCRIPTION
**Pull request by LuciferYang**
### What changes were proposed in this pull request?
This pr  add support to make `recentProgress` and `lastProgress` in `RemoteStreamingQuery` return `StreamingQueryProgress` instance consistent with the native Scala Api.

### Why are the changes needed?
Add Spark connect jvm client api coverage.

### Does this PR introduce _any_ user-facing change?
No

### How was this patch tested?
Add new check to `StreamingQuerySuite`

### CODE REVIEW

File: connector/connect/client/jvm/src/main/scala/org/apache/spark/sql/streaming/progress.scala [hunk 1]
```diff
@@ -17,6 +17,297 @@
...
+    val batchId: Long,
+    val batchDuration: Long,
+    val durationMs: ju.Map[String, JLong],
+    val eventTime: ju.Map[String, String],
+    val stateOperators: Array[StateOperatorProgress],
+    val sources: Array[SourceProgress],
+    val sink: SinkProgress,
+    @JsonSetter(nulls = Nulls.AS_EMPTY)
+    @JsonDeserialize(contentAs = classOf[GenericRowWithSchema])
+    val observedMetrics: ju.Map[String, Row])
```
Comment (LuciferYang) [inline connector/connect/client/jvm/src/main/scala/org/apache/spark/sql/streaming/progress.scala:158]: `observedMetrics` maybe null, so add `@JsonSetter(nulls = Nulls.AS_EMPTY)` to set it as empty map when it is null

File: connector/connect/client/jvm/src/main/scala/org/apache/spark/sql/streaming/progress.scala [hunk 2]
```diff
@@ -17,6 +17,297 @@
...
+    val numRowsUpdated: Long,
+    val allUpdatesTimeMs: Long,
+    val numRowsRemoved: Long,
+    val allRemovalsTimeMs: Long,
+    val commitTimeMs: Long,
+    val memoryUsedBytes: Long,
+    val numRowsDroppedByWatermark: Long,
+    val numShufflePartitions: Long,
+    val numStateStoreInstances: Long,
+    val customMetrics: ju.Map[String, JLong] = new ju.HashMap())
```
Comment (LuciferYang) [inline connector/connect/client/jvm/src/main/scala/org/apache/spark/sql/streaming/progress.scala:56]: Should we also add `@JsonSetter(nulls = Nulls.AS_EMPTY)` as protection for other collection types?

File: connector/connect/client/jvm/src/main/scala/org/apache/spark/sql/streaming/progress.scala [hunk 3]
```diff
@@ -17,6 +17,297 @@
...
+import org.apache.spark.sql.Row
+import org.apache.spark.sql.catalyst.expressions.GenericRowWithSchema
+import org.apache.spark.sql.streaming.SafeJsonSerializer.{safeDoubleToJValue, safeMapToJValue}
+import org.apache.spark.sql.streaming.SinkProgress.DEFAULT_NUM_OUTPUT_ROWS
+
+/**
+ * Information about updates made to stateful operators in a [[StreamingQuery]] during a trigger.
+ */
+@Evolving
+class StateOperatorProgress private[spark] (
```
Comment (LuciferYang) [inline connector/connect/client/jvm/src/main/scala/org/apache/spark/sql/streaming/progress.scala:44]: Just used in `[spark]` , So no mima check was added

**Comment by LuciferYang** [comment]
If this is not consistent with the initial idea of this ticket, please let me know

**Comment by HyukjinKwon** [comment]
cc @HeartSaVioR

File: connector/connect/client/jvm/src/main/scala/org/apache/spark/sql/streaming/progress.scala [hunk 4]
```diff
@@ -17,6 +17,297 @@
...
+    val stateOperators: Array[StateOperatorProgress],
+    val sources: Array[SourceProgress],
+    val sink: SinkProgress,
+    @JsonSetter(nulls = Nulls.AS_EMPTY)
+    @JsonDeserialize(contentAs = classOf[GenericRowWithSchema])
+    val observedMetrics: ju.Map[String, Row])
+    extends Serializable {
+
+  /** The aggregate (across all sources) number of records processed in a trigger. */
+  def numInputRows: Long = sources.map(_.numInputRows).sum
```
Comment (bogao007) [inline connector/connect/client/jvm/src/main/scala/org/apache/spark/sql/streaming/progress.scala:162]: Aggregated metrics like `numInputRows`, `inputRowsPerSecond` and `processedRowsPerSecond` are being calculated inside this class. Can they be safely deserialized from json string? Might also worth to add them into the test case.

**Comment by bogao007** [comment]
cc @rangadi

File: connector/connect/client/jvm/src/main/scala/org/apache/spark/sql/streaming/progress.scala (hunk 1 above)
Comment (LuciferYang) [reply connector/connect/client/jvm/src/main/scala/org/apache/spark/sql/streaming/progress.scala:158]: Deserialize result of `observedMetrics` has problem, `scheam` fo GenericRowWithSchema is null

Comment (LuciferYang) [reply connector/connect/client/jvm/src/main/scala/org/apache/spark/sql/streaming/progress.scala:158]: https://github.com/apache/spark/blob/d407a42090d7c027050be7ee723f7e3d8f686ed7/connector/connect/server/src/main/scala/org/apache/spark/sql/connect/planner/SparkConnectPlanner.scala#L2182-L2186

@rangadi @bogao007 I think we should serialize the `StreamingQueryProgress` to json as a pojo to  maintain the row schema information in `observedMetrics`, `progressReports.map(_.json)` will lose schmea information of `GenericRowWithSchema`, this results in the inability to deserialize to GenericRowWithSchema on the client side

Comment (LuciferYang) [reply connector/connect/client/jvm/src/main/scala/org/apache/spark/sql/streaming/progress.scala:158]: Maybe py test will fail, let's wait for ci

Comment (bogao007) [reply connector/connect/client/jvm/src/main/scala/org/apache/spark/sql/streaming/progress.scala:158]: Yep, we would not be able to deserialize `observedMetrics` without this change, thanks for the update!

**Review by bogao007 — APPROVED** [review]
LGTM overall, with one minor comment on the test case. Please also get an approval from @rangadi before merging, thanks!

File: connector/connect/client/jvm/src/test/scala/org/apache/spark/sql/streaming/StreamingQueryProgressSuite.scala [hunk 5]
```diff
@@ -0,0 +1,214 @@
...
+         |}
+      """.stripMargin.trim
+
+    // scalastyle:off
+    println(jsonStringFromServerSide)
+
+    val result = StreamingQueryProgress.fromJson(jsonStringFromServerSide)
+    assert(result.id.toString === "33ac26f4-1c39-46ce-b798-f3d2a21211d4")
+    assert(result.runId.toString === "849c2c9a-b9f8-446f-9180-259a60fd888c")
+    assert(result.numInputRows === 678)
```
[the remark this answers is not in the document]
Comment (LuciferYang) [reply connector/connect/client/jvm/src/test/scala/org/apache/spark/sql/streaming/StreamingQueryProgressSuite.scala]: Added a `sources` item, there change to `assert(result.numInputRows === 1467) // 678 + 789`

File: connector/connect/client/jvm/src/main/scala/org/apache/spark/sql/streaming/StreamingQuery.scala [hunk 6]
```diff
@@ -163,14 +163,14 @@ class RemoteStreamingQuery(
...
-      .map(json => new StreamingQueryProgress(json))
+      .map(StreamingQueryProgress.fromJson)
       .toArray
   }
 
   override def lastProgress: StreamingQueryProgress = {
     executeQueryCmd(
       _.setLastProgress(true)).getRecentProgress.getRecentProgressJsonList.asScala.headOption
-      .map(json => new StreamingQueryProgress(json))
+      .map(StreamingQueryProgress.fromJson)
```
Comment (rangadi) [inline connector/connect/client/jvm/src/main/scala/org/apache/spark/sql/streaming/StreamingQuery.scala:173]: This is a good idea (at least for now). So essentially we don't need to add many protobuf message to match :).

**Review by rangadi — APPROVED** [review]
Thanks for adding this!

**Review by HyukjinKwon — APPROVED** [review]

**Comment by LuciferYang** [comment]
rebased and resolved conflicts

**Comment by LuciferYang** [comment]
friendly ping @HyukjinKwon can we merge this one?

**Comment by HeartSaVioR** [comment]
I'll leave this to you for self-merging, so that you can test your new permission. Congrats again :)

**Comment by LuciferYang** [comment]
> I'll leave this to you for self-merging, so that you can test your new permission. Congrats again :)

Thanks @HeartSaVioR :)

**Comment by HeartSaVioR** [comment]
I'll just help merging this one as it has been here for multiple weeks and we don't want to require this PR to be rebased anymore.

**Comment by HeartSaVioR** [comment]
Thanks @LuciferYang , I merged this to master.

**Pull request closed without merging**